server/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -263,76 +263,39 @@ private static <V> V removeMatch(Map<Set<RelationName>, V> valuesByNames, Set<Re
      * @return the new list of {@link JoinPair}
      */
     static List<JoinPair> convertImplicitJoinConditionsToJoinPairs(List<JoinPair> explicitJoinPairs,
-                                                                   Map<Set<RelationName>, Symbol> splitQueries) {
-        Iterator<Map.Entry<Set<RelationName>, Symbol>> queryIterator = splitQueries.entrySet().iterator();
-        ArrayList<JoinPair> newJoinPairs = new ArrayList<>(explicitJoinPairs.size() + splitQueries.size());
-        newJoinPairs.addAll(explicitJoinPairs);
+                                                                   Map<Set<RelationName>, Symbol> implicitJoinConditions) {
+        ArrayList<JoinPair> newJoinPairs = new ArrayList<>(explicitJoinPairs);
+        Iterator<Map.Entry<Set<RelationName>, Symbol>> queryIterator = implicitJoinConditions.entrySet().iterator();
 
         while (queryIterator.hasNext()) {
             Map.Entry<Set<RelationName>, Symbol> queryEntry = queryIterator.next();
             Set<RelationName> relations = queryEntry.getKey();
-
-            if (relations.size() == 2) { // If more than 2 relations are involved it cannot be converted to a JoinPair
+            // If more than 2 relations are involved it cannot be converted to a JoinPair
+            if (relations.size() == 2) {
                 Symbol implicitJoinCondition = queryEntry.getValue();
-                JoinPair newJoinPair = null;
-                int existingJoinPairIdx = -1;
-                for (int i = 0; i < explicitJoinPairs.size(); i++) {
-                    JoinPair joinPair = explicitJoinPairs.get(i);
-                    if (relations.contains(joinPair.left()) && relations.contains(joinPair.right())) {
-                        existingJoinPairIdx = i;
-                        // If a JoinPair with the involved relations already exists then depending on the JoinType:
-                        //  - INNER JOIN:  the implicit join condition can be "AND joined" with
-                        //                 the existing explicit condition of the JoinPair.
-                        //
-                        //  - CROSS JOIN:  the implicit join condition becomes explicit and set on the JoinPair
-                        //                 which becomes an INNER JOIN
-                        //
-                        //  - Outer Joins: Queries in the WHERE clause must be semantically applied after the outer join
-                        //                 and cannot be merged with the conditions of the ON clause.
-                        //
-                        //  - SEMI/ANTI:   Queries in the WHERE clause must be semantically applied after the SEMI/ANTI
-                        //                 join and cannot be merged with the generated SEMI/ANTI condition.
-                        //
-                        if (joinPair.joinType() == JoinType.INNER || joinPair.joinType() == JoinType.CROSS) {
```

Comment:
The filter-push-down logic in the optimizer already covers the inner-join logic. The cross-join logic is moved into a own rule and the semi/anti join behaviour logic is just a note.